packages/@lwc/engine/src/framework/base-lightning-element.ts

```diff
@@ -341,19 +341,24 @@ BaseLightningElementConstructor.prototype = {
         const wrappedListener = getWrappedComponentsListener(vm, listener);
         vm.elm.removeEventListener(type, wrappedListener, options);
     },
-    setAttributeNS(ns: string | null, attrName: string, _value: string) {
+    hasAttribute(): string | null {
```

Comment:
learned something new recently, you can reuse the types from another method by using something like this:

```ts
hasAttribute(this: LightningElement, ...[attr]: Parameters<typeof Element.prototype.hasAttribute>): ReturnType<typeof Element.prototype.hasAttribute> {
        const elm = getElementFrom(this);
        return elm.hasAttribute.apply(elm, arguments);
    }
```

what this does is that it will use the type declaration for arguments and return value from `Element.prototype.hasAttribute`, so you don't have to duplicate that here.

Reply:
more info about this here: https://2ality.com/2020/04/typing-functions-typescript.html#checking-function-declarations-(extravagant)

Reply:
Instead of extracting the method's parameters and return types, I would rather prefer using the `Pick` typescript operator to create a subset of the `Element` interfaces and make the base class implement this interface. This would avoid the repetition: [playground](https://www.typescriptlang.org/play/index.html?ssl=17&ssc=2&pln=1&pc=1#code/C4TwDgpgBAogNhAthAdsAKgCwgcQPZ4AmACgIYBOwAzlALxTECWAxgNYA88SqwANFAHIA5hGABBYMHKMARgFdgEAVAA+gqqIlTZCpasGZSVLdPmKBAPgDcAKBvM4RmgBlGQzMBSMUQrsjRQjIhgCP7UsKE8WLgEJBThAN42UClQzHgoVFJyzMB45AAUyaklYNIAbqSKUBBwiABcEdxoxSkAlFAJUAC+diUi4pKmugUopMiNWdI+HV2tJVDkonLkKFDAmIxUAHS1iNsDJjqKo+MQbbYlvSWGxkPHEKcTUFPeQrPzJUvAK2sbW7s6ttbkczI8xsgLvNrqkNINtGCnhBJtofPxKnA5MiXqj3p0oJ9Ut9futNjs9ts4aCRhCIOjSJjzpdUr1ukA)

```ts
type ElementTheGoodParts = Pick<Element, 'getAttribute' | 'setAttribute' | 'hasAttribute'>;

class LightningElement implements ElementTheGoodParts {
    constructor(
        private elm: Element
    ) { }

    getAttribute(name: string) { 
        return this.elm.getAttribute(name);
    }
    hasAttribute(name: string) {
        return this.elm.hasAttribute(name);
    }
    setAttribute(name: string, value: string) { 
        return this.elm.setAttribute(name, value);
    }
}
```

Reply:
Will address types in another PR